packages/notifications/src/error.ts

```diff
@@ -0,0 +1,87 @@
+import { AnyPromiseFn } from '@terra-ui-packages/core-utils';
+
+import { NotificationsContract, Notifier } from './useNotifications';
+
+export type ErrorCallback = (error: unknown) => void | Promise<void>;
+
+/**
+ * Should return true of a given error should be ignored
+ */
+export type IgnoreErrorDecider = (title: string, obj?: unknown) => boolean;
+
+/**
+ * Invoke the `callback` with any error thrown when evaluating the async `fn` with `...args`.
+ */
+export const withErrorHandling = (callback: ErrorCallback) => {
+  return <F extends AnyPromiseFn>(fn: F) => {
+    return (async (...args: Parameters<F>) => {
+      try {
+        return await fn(...args);
+      } catch (error) {
+        await callback(error);
+      }
+    }) as F;
+  };
+};
+
+/**
+ * Return a Promise to the result of evaluating the async `fn` with `...args` or undefined if
+ * evaluation fails.
+ */
+export const withErrorIgnoring = withErrorHandling(() => {});
+
+/**
+ * allows composition of NotificationsContract dependency when useful for code reuse
+ * @param notifier
+ */
+export const makeNotificationsContract = (
+  notifier: Notifier,
+  ignoreError: IgnoreErrorDecider
+): NotificationsContract => ({
+  notify: notifier.notify,
+
+  reportError: async (title: string, obj?: unknown) => {
+    console.error(title, obj); // helpful when the notify component fails to render
+    // Do not show an error notification if error should be ignored
+    if (ignoreError(title, obj)) {
+      return;
+    }
+    const detail = await (obj instanceof Response ? obj.text().catch(() => 'Unknown error') : obj);
+    notifier.notify('error', title, { detail });
+  },
+
+  /**
+   * Return a Promise to the result of evaluating the async `fn` with `...args`. If evaluation fails,
+   * report the error to the user with `title` as a side effect.
+   */
+  reportErrorAndRethrow: (title: string) => {
+    return <F extends AnyPromiseFn>(fn: F) =>
+      withErrorHandling(async (error): Promise<void> => {
+        await makeNotificationsContract(notifier, ignoreError).reportError(title, error);
+        throw error;
+      })(fn);
+  },
+
+  /**
+   *  This function is designed for use in modals.
+   *  Modals can overlay any error reporting, with the `throw` in the default `withErrorReporting`
+   *  preventing the modal itself from closing on error
+   *  As such, we must ensure we call the dismiss function if an error occurs
+   */
+  withErrorReportingInModal: (title: string, onDismiss: () => void) => {
+    return withErrorHandling(async (error) => {
+      await makeNotificationsContract(notifier, ignoreError).reportError(title, error);
+      onDismiss();
+      throw error;
+    });
+  },
```

Comment:
Not too keen on concerning this package with modals. I think this could be defined in Terra UI.

Also, generally, I think this function encourages lazy error handling and would vote against promoting it to a package. Often a modal involves submitting some form and 400 errors should be handled by showing which field was invalid and allowing the user to adjust instead of closing the modal, etc.

Reply:
I think I'd still vote for getting the "modal" specific names out of it. This is more generic... the default error reporting + some onError callback.

Reply:
hmmm... yeah, the variants are a bit odd if you look a them collectively.... including reportErrorAndRethrow should probably be named withErrorReportingandRetrow since it returns a function agumenter.

what about simplifying like this:

```
export interface ErrorReportingOptions {
  rethrow?: boolean;
  onNotified?: () => void;
}

export interface NotificationsContract extends Notifier {
  reportError: (title: string, obj?: unknown) => Promise<void>;

  /**
   * Returns a function augmenter (HoF).
   * If provided function fails when called, report the error to the user with `title` as a side effect.
   * Use options arg to give onNotified callback, and/or rethrow on error.
   */
  withErrorReporting: <F extends AnyPromiseFn>(title: string, options?: ErrorReportingOptions) => (fn: F) => F;
}
```

Reply:
Sounds good. A few naming nits to avoid leaking "notifications" out of "error reporting"...

```ts
export interface ErrorReportingOptions {
  rethrow?: boolean; // default false
  onReport?: () => void;
}

export interface ErrorReporter extends Notifier {
  reportError: (title: string, obj?: unknown) => Promise<void>;

  /**
   * Returns a function augmenter (HoF).
   * If provided function fails when called, report the error to the user with `title` as a side effect.
   * Use options arg to give onNotified callback, and/or rethrow on error.
   */
  withErrorReporting: <F extends AnyPromiseFn>(title: string, options?: ErrorReportingOptions) => (fn: F) => F;
}
```

Reply:
ok cool, tho this last point blends into your other comment.... the tricky bit is that Environments (and similar large components) have components and hooks that need both Notifier.notify and the methods we are settling on here for "ErrorReporter".. This PR is blending them together into the NotificationsContract so that we have just one "thing" to worry about in context.... trying to avoid context explosion... thoughts?

Reply:
My thinking is that "error reporting" is a thing, then it's separate from (though build on) the base "notifications".
```
type NotificationsContract = Notifier & ErrorReporter;
```

Or, if "error reporting" is part of "notifications", then "error reporting" is not a thing on its own and does not appear in types.
```ts
type Notifications extends BaseNotifications {
  reportError: ...
  ...
}
```